service/history/workflow_task_handler_callbacks.go

```diff
@@ -809,6 +808,10 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 		return nil, updateErr
 	}
 
+	// If mutable state was persisted successfully (or persistence was skipped),
+	// then effects needs to be applied immediately to keep registry and mutable state in sync.
+	effects.Apply(ctx)
```

Comment:
The fix is to move this line from `defer` to here.

Reply:
Is it possible to write a unit test that repros this?

Reply:
I was thinking about functional test and it is not possible w/o changing functional test framework to use fault injections. Unit test should be possible.

Reply:
I don't really agree with this :) In my opinion:

A bug fix should always be accompanied by a test reproducing the bug and preventing regression if that test is not too much work to produce.

> I don't want to add just one test for this specific case. I want to cover success cases too.

There are always more tests to write tomorrow. Adding one test to reproduce is bug, even when the rest of the file is uncovered, is normal and good practice.